src/main/java/com/squareup/javapoet/CodeBlock.java

```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
         if (format.charAt(p) != '$') {
           nextP = format.indexOf('$', p + 1);
           if (nextP == -1) nextP = format.length();
+          formatParts.add(format.substring(p, nextP));
         } else {
           checkState(p + 1 < format.length(), "dangling $ in format string %s", format);
-          switch (format.charAt(p + 1)) {
+          /* Rules:
+           * Format must be in the form $[Numeric Index]{N, L, S, T}
+           * OR
+           * in the form ${$, >, <, [, ]}
+           * Numeric index is ONE BASED (for consistency with java formatting)
+           * Numeric indexing must refer to a format argument within the args array
+           * Only progress iteration over the argument iterator if we do not hit an indexed entry
+           */
+          int countOfIndexCharacters = 0;
+          while (Character.isDigit(format.charAt(p + countOfIndexCharacters + 1))) {
+            countOfIndexCharacters++;
+          }
+          int argsIndex = -1;
+          if (countOfIndexCharacters != 0) {
+            argsIndex = Integer.parseInt(format.substring(p + 1, p + 1 + countOfIndexCharacters));
+            checkArgument(!(argsIndex > args.length),
+                "Argument index %s in '%s' is larger than number of parameters",
+                argsIndex, format);
+          }
+          switch (format.charAt(p + countOfIndexCharacters + 1)) {
             case 'N':
-              this.args.add(argToName(i.next()));
+              this.args.add(argToName((argsIndex == -1) ? i.next() : args[argsIndex - 1]));
               break;
             case 'L':
-              this.args.add(argToLiteral(i.next()));
+              this.args.add(argToLiteral((argsIndex == -1) ? i.next() : args[argsIndex - 1]));
               break;
             case 'S':
-              this.args.add(argToString(i.next()));
+              this.args.add(argToString((argsIndex == -1) ? i.next() : args[argsIndex - 1]));
               break;
             case 'T':
-              this.args.add(argToType(i.next()));
+              this.args.add(argToType((argsIndex == -1) ? i.next() : args[argsIndex - 1]));
               break;
             case '$':
             case '>':
             case '<':
             case '[':
             case ']':
+              checkState(argsIndex == -1, "$$, $>, $<, $[ and $] may not have an index");
               break;
             default:
               throw new IllegalArgumentException("invalid format string: " + format);
           }
 
-          nextP = p + 2;
+          nextP = p + countOfIndexCharacters + 2;
+          formatParts.add(countOfIndexCharacters == 0
```

Comment:
Does it do the right thing automatically without the ternary? I think it might just work